crates/flipperzero/src/subghz/device.rs

```diff
@@ -0,0 +1,193 @@
+        let actual_freq = unsafe { subghz_devices_set_frequency(self.device, freq_u32) };
+
+        if actual_freq != 0 {
+            Ok(Frequency::new::<hertz>(actual_freq))
+        } else {
+            Err(SubGhzError::UnableToSetFrequency)
+        }
+    }
+
+    pub fn is_frequency_valid(&mut self, freq: Frequency) -> bool {
+        unsafe { subghz_devices_is_frequency_valid(self.device, freq.value) }
+    }
+
+    pub fn set_async_mirror_pin(&mut self) {
+        unimplemented!("Should probably create safe abstraction around GpioPin first, have not investigated if this is needed");
+    }
+
+    pub fn get_data_gpio(&mut self) {
+        unimplemented!("Should probably create safe abstraction around GpioPin first, have not investigated if this is needed");
+    }
+
+    pub fn set_tx(&mut self) -> Result<(), SubGhzError> {
+        let result = unsafe { subghz_devices_set_tx(self.device) };
+
+        // Returns true if Ok
+        if result {
+            Ok(())
+        } else {
+            Err(SubGhzError::CannotTxOnFrequency)
+        }
+    }
+
+    pub fn flush_tx(&mut self) {
+        unsafe { subghz_devices_flush_tx(self.device) };
+    }
+
+    pub fn set_rx(&mut self) {
+        unsafe {
+            subghz_devices_set_rx(self.device);
+        }
+    }
+
+    pub fn flush_rx(&mut self) {
+        unsafe {
+            subghz_devices_flush_rx(self.device);
+        }
+    }
+
+    pub fn get_rssi(&mut self) -> Option<f32> {
+        let rssi_function = unsafe { self.device.interconnect.as_ref() }?.get_rssi?;
+        Some(unsafe { rssi_function() })
+    }
+
+    pub fn get_lqi(&mut self) -> Option<u8> {
+        let lqi_function = unsafe { self.device.interconnect.as_ref()?.get_lqi? };
+        Some(unsafe { lqi_function() })
+    }
+
+    pub fn rx_pipe_not_empty(&mut self) -> bool {
+        unsafe { subghz_devices_rx_pipe_not_empty(self.device) }
+    }
+
+    pub fn is_rx_crc_valid(&mut self) -> bool {
+        unsafe { subghz_devices_is_rx_data_crc_valid(self.device) }
+    }
+
+    pub fn read_packet(&mut self, buf: &mut [u8]) -> usize {
+        let mut size = if buf.len() > 255 {
+            255_u8
+        } else {
+            buf.len() as u8
+        };
+
+        unsafe {
+            subghz_devices_read_packet(self.device, buf.as_mut_ptr(), &mut size);
+        };
+        size as usize
+    }
+
+    pub fn write_packet(&mut self, buf: &[u8]) -> Result<(), SubGhzError> {
+        if buf.len() > 255 {
+            return Err(SubGhzError::PacketTooLong);
+        }
+
+        let size = buf.len() as u8;
+
+        unsafe {
+            subghz_devices_write_packet(self.device, buf.as_ptr(), size);
+        }
+        Ok(())
+    }
+}
+
+impl Drop for SubGhz {
+    fn drop(&mut self) {
+        // Safety: self.device is not Null so this will not crash when furi_check is invoked.
+        unsafe {
+            subghz_devices_end(self.device);
+            subghz_devices_deinit();
```

Comment:
What happens if multiple instances of `SubGhz` are created and thus `subghz_devices_deinit` is called (on ones destructor) while the other one is still active?

I guess we may need to address this upstream, e.g. request adding a counter to it (e.g. the number of currently active *activations*)

Reply:
In some haphazard test code:

```rust
fn main(_args: Option<&CStr>) -> i32 {
    let dev_name = c"cc1101_int";

    let Some(mut x) = subghz::SubGhz::subghz_devices_get_by_name(dev_name) else {
        alert("BOOOO");
        return 0;
    };

    x.load_preset(subghz::preset::SubGhzPreset::Ook270Async);

    let Ok(()) = x.begin() else {
        return 0;
    };

    let Ok(actual_freq) = x.set_frequency(Frequency::new::<hertz>(433_500_000)) else {
        return 0;
    };

    let mut s = String::<128>::new();

    alert("about to take second ref");
    // Crash
    let Some(y) = subghz::SubGhz::subghz_devices_get_by_name(dev_name) else {
        alert("Cant take second reference");
        return 0;
    };

    0
}
```

Attempting to create a second reference crashes the device (I am not too surprised since the API does a lot of checking for safety). That being said, I don't know which function actually crashes things, so I don't know how it is enforced: I would like to be able to enforce this to prevent the crash.